src/Native/System.Private.CoreLib.Native/pal_time.cpp

```diff
@@ -33,3 +34,161 @@ mach_timebase_info_data_t *InitializeTimebaseInfo()
     return g_isMachTimebaseInfoInitialized ? &g_machTimebaseInfo : nullptr;
 }
 #endif
+
+extern "C" uint64_t CoreLibNative_GetHighPrecisionCounts()
+{
+    uint64_t counts = 0;
+
+#if HAVE_MACH_ABSOLUTE_TIME
+    {
+        counts = mach_absolute_time();
+    }
+#elif HAVE_CLOCK_MONOTONIC_COARSE || HAVE_CLOCK_MONOTONIC
+    {
+        clockid_t clockType =
+#if HAVE_CLOCK_MONOTONIC_COARSE
+            CLOCK_MONOTONIC_COARSE; // good enough resolution, fastest speed
+#else
+            CLOCK_MONOTONIC;
+#endif
+        struct timespec ts;
+        if (clock_gettime(clockType, &ts) != 0)
+        {
+            assert(false);
+            return counts;
+        }
+        counts = ts.tv_sec * NanosecondsPerSecond + ts.tv_nsec;
+    }
+#else
+    {
+        struct timeval tv;
+        if (gettimeofday(&tv, nullptr) == -1)
+        {
+            assert(false);
+            return counts;
+        }
+        counts = tv.tv_sec * MicrosecondsPerSecond + tv.tv_usec;
+    }
+#endif
+    return counts;
+}
+
+extern "C" uint64_t CoreLibNative_GetHighPrecisionCounterFrequency()
+{
+    static uint64_t frequency = 0;
+    if (frequency != 0)
+    {
+        return frequency;
+    }
+
+#if HAVE_MACH_ABSOLUTE_TIME
+    {
+        mach_timebase_info_data_t *machTimebaseInfo = GetMachTimebaseInfo();
+        frequency = NanosecondsPerSecond * static_cast<uint64_t>(machTimebaseInfo->denom) / machTimebaseInfo->numer;
+    }
+#elif HAVE_CLOCK_MONOTONIC_COARSE || HAVE_CLOCK_MONOTONIC
+    {
+        frequency = NanosecondsPerSecond;
+    }
+#else
+    {
+        frequency = MicrosecondsPerSecond;
+    }
+#endif
+    return frequency;
+}
+
+// Returns a 64-bit tick count with a millisecond resolution. It tries its best
+// to return monotonically increasing counts and avoid being affected by changes
+// to the system clock (either due to drift or due to explicit changes to system
+// time).
+extern "C" uint64_t CoreLibNative_GetTickCount64()
```

Comment:
This function tends to be performance sensitive. This change seems to be making it slower than what it needs to be...